src/natcap/invest/scenic_quality/viewshed.pyx

```diff
@@ -638,22 +639,21 @@ def viewshed(dem_raster_path_band,
         # than or equal to the minimum-visible height AND is closer than the
         # maximum visible radius.
         target_dem_height = dem_managed_raster.get(n, m)
-        adjusted_dem_height = dem_managed_raster.get(n, m) - adjustment
-        if (adjusted_dem_height >= z and
-                target_pixel.distance_to_viewpoint < max_visible_radius and
-                target_dem_height != nodata):
-            visibility_managed_raster.set(n, m, 1)
+        adjusted_dem_height = target_dem_height - adjustment
+
+        # If it's close enough to nodata to be interpreted as nodata,
+        # consider it to be nodata.  Nodata implies that visibility is
+        # undefined ... which it is, since there's no defined DEM value for
+        # this pixel.
+        if is_close(target_dem_height, nodata):
+            visibility_managed_raster.set(n, m, VISIBILITY_NODATA)
+            aux_managed_raster.set(n, m, z)
+        elif (adjusted_dem_height >= z and
+                target_pixel.distance_to_viewpoint < max_visible_radius):
+            visibility_managed_raster.set(n, m, 1)  # the pixel is visible
             aux_managed_raster.set(n, m, adjusted_dem_height)
         else:
-            # If it's close enough to nodata to be interpreted as nodata,
-            # consider it to be nodata.  Nodata implies that visibility is
-            # undefined ... which it is, since there's no defined DEM value for
-            # this pixel.
-            if math.fabs(target_dem_height - nodata) <= 1.0e-7:

```

Comment:
The logic of this conditional was confusing to me and had a potential bug: if `adjusted_dem_height >= z` AND `target_pixel.distance_to_viewpoint < max_visible_radius` AND `target_dem_height != nodata`, but is within the float tolerance of `nodata`, then the first condition would apply. But if `target_dem_height` exactly equaled `nodata`, then the second condition would apply. So this change could produce different results in that case, but I think this is what was intended.